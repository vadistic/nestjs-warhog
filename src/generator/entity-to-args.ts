import { EntityMetadata } from 'typeorm'

import { niceString } from '../utils'

// FIND

export const entityToFindOneArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}WhereUniqueArgs {
      @Field(() => ${entity.name}WhereInput)
      where!: ${entity.name}WhereUniqueInput
    }
  `
}

// FIND MANY

export const entityToFindArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}WhereArgs extends PaginationArgs {
      @Field(() => ${entity.name}WhereInput, { nullable: true })
      where?: ${entity.name}WhereInput
      @Field(() => ${entity.name}OrderByEnum, { nullable: true })
      orderBy?: ${entity.name}OrderByEnum
    }
  `
}

// CREATE

export const entityToCreateArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}CreateArgs {
      @Field(() => ${entity.name}CreateInput)
      data!: ${entity.name}CreateInput
    }
  `
}

// CREATE MANY

export const entityToCreateManyArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}CreateManyArgs {
      @Field(() => [${entity.name}CreateInput])
      data!: ${entity.name}CreateInput[]
    }
  `
}

// UPDATE

export const entityToUpdateArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}UpdateArgs {
      @Field(() => ${entity.name}UpdateInput)
      data!: ${entity.name}UpdateInput
      @Field(() => ${entity.name}WhereUniqueInput)
      where!: ${entity.name}WhereUniqueInput
    }
  `
}

// UPDATE MANY

export const entityToUpdateManyArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}UpdateManyArgs {
      @Field(() => ${entity.name}UpdateInput)
      data!: ${entity.name}UpdateInput
      @Field(() => ${entity.name}WhereInput)
      where!: ${entity.name}WhereInput
    }
  `
}

// DELETE

export const entityToDeleteArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}DeleteArgs {
      @Field(() => ${entity.name}WhereUniqueInput)
      where!: ${entity.name}WhereUniqueInput
    }
  `
}

// DELETE MANY

export const entityToDeleteManyArgs = (entity: EntityMetadata): string => {
  return niceString`
    @ArgsType()
    export class ${entity.name}DeleteManyArgs {
      @Field(() => ${entity.name}WhereInput)
      where!: ${entity.name}WhereInput
    }
  `
}
