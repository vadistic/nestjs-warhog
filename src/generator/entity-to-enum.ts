import { EntityMetadata } from 'typeorm/metadata/EntityMetadata'
import { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata'

import { niceString, indent } from '../utils'

// ORDER BY

export const entityToOrderByEnum = (entity: EntityMetadata): string => {
  let fieldsTemplate = ''

  entity.columns.forEach((column: ColumnMetadata) => {
    if (!column.isPrimary && !column.isVersion) {
      fieldsTemplate += `${column.propertyName}_ASC = '${column.propertyName}_ASC',\n`
      fieldsTemplate += `${column.propertyName}_DESC = '${column.propertyName}_DESC',\n`
    }
  })

  return niceString`
    export enum ${entity.name}OrderByEnum {
    ${indent(fieldsTemplate)}
    }

    registerEnumType(${entity.name}OrderByEnum, {
      name: '${entity.name}OrderByInput'
    })
  `
}
