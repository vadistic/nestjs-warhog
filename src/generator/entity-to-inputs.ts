import { EntityMetadata } from 'typeorm'
import { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata'
import { GraphQLID, GraphQLString, GraphQLFloat, GraphQLInt } from 'graphql'
import { GraphQLISODateTime } from 'type-graphql'

import {
  columnTypeToGraphQLDataType,
  columnToTypeScriptType,
  uniquesForEntity,
  columnToTypes,
} from './typeorm-converter'
import { niceString, indent } from '../utils'

const CREATE_UPDATE_BLACKLIST = [
  'createdAt',
  'createdById',
  'updatedAt',
  'updatedById',
  'deletedAt',
  'deletedById',
]

// WHERE UNIQUE

export const entityToWhereUniqueInput = (entity: EntityMetadata): string => {
  const uniques = uniquesForEntity(entity)

  const numUniques = entity.columns.reduce<number>((num, column: ColumnMetadata) => {
    if (uniques.includes(column.propertyName) || column.isPrimary) {
      num++
    }

    return num
  }, 0)

  // If there is only one unique field, it should not be nullable
  const uniqueFieldsAreNullable = numUniques > 1

  let fieldsTemplate = ''

  entity.columns.forEach((column: ColumnMetadata) => {
    if (uniques.includes(column.propertyName) || column.isPrimary) {
      const nullable = uniqueFieldsAreNullable ? ', { nullable: true }' : ''
      const graphQLDataType = columnTypeToGraphQLDataType(column)
      const tsType = columnToTypeScriptType(column)

      fieldsTemplate += niceString`
        @Field(() => ${graphQLDataType}${nullable})
        ${column.propertyName}?: ${tsType}
      `
    }
  })

  return niceString`
    @InputType()
    export class ${entity.name}WhereUniqueInput {
    ${indent(fieldsTemplate)}
    }
  `
}

// WHERE

export const entityToWhereInput = (entity: EntityMetadata): string => {
  let fieldTemplates = ''

  entity.columns.forEach((column: ColumnMetadata) => {
    const { graphqlType, tsType } = columnToTypes(column)
    const graphQLDataType = columnTypeToGraphQLDataType(column)

    // TODO: for foreign key fields, only allow the same filters as ID below
    // Example: photo.userId: String
    if (column.isPrimary || graphqlType === GraphQLID) {
      fieldTemplates += niceString`
        @Field(() => ${graphQLDataType},{ nullable: true })
        ${column.propertyName}_eq?: string
        @Field(() => [${graphQLDataType}], { nullable: true })
        ${column.propertyName}_in?: string[]
      `
    } else if (graphqlType === GraphQLString) {
      fieldTemplates += niceString`
        @Field({ nullable: true })
        ${column.propertyName}_eq?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_not?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_contains?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_startsWith?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_endsWith?: ${tsType}
        @Field(() => [${graphQLDataType}], { nullable: true })
        ${column.propertyName}_in?: ${tsType}[]
      `
    } else if (graphqlType === GraphQLFloat || graphqlType === GraphQLInt) {
      fieldTemplates += niceString`
        @Field({ nullable: true })
        ${column.propertyName}_eq?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_gt?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_gte?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_lt?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_lte?: ${tsType}
        @Field(() => [${graphQLDataType}], { nullable: true })
        ${column.propertyName}_in?: ${tsType}[]
      `
    } else if (graphqlType === GraphQLISODateTime) {
      fieldTemplates += niceString`
        @Field({ nullable: true })
        ${column.propertyName}_gt?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_gte?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_lt?: ${tsType}
        @Field({ nullable: true })
        ${column.propertyName}_lte?: ${tsType}
      `
    } else if (
      column.type !== 'json' &&
      column.type !== 'jsonb' &&
      column.type !== 'varying character'
    ) {
      // TODO: JSON??

      fieldTemplates += niceString`
        @Field(() => ${graphQLDataType}, { nullable: true })
        ${column.propertyName}_eq?: ${graphQLDataType}
        @Field(() => [${graphQLDataType}], { nullable: true })
        ${column.propertyName}_in?: ${graphQLDataType}[]
      `
    }

    fieldTemplates += '\n\n'
  })

  return niceString`
    @InputType()
    export class ${entity.name}WhereInput {
    ${indent(fieldTemplates)}
    }
  `
}

// CREATE

export const entityToCreateInput = (entity: EntityMetadata): string => {
  let fieldTemplates = ''

  entity.columns.forEach((column: ColumnMetadata) => {
    if (
      !column.isPrimary &&
      !column.isCreateDate &&
      !column.isGenerated &&
      !column.isUpdateDate &&
      !column.isVersion &&
      !CREATE_UPDATE_BLACKLIST.includes(column.propertyName)
    ) {
      const graphQLDataType = columnTypeToGraphQLDataType(column)
      const nullable = column.isNullable ? '{ nullable: true }' : ''
      const tsRequired = column.isNullable ? '?' : '!'
      const tsType = columnToTypeScriptType(column)

      // we need to know what the graphql type is and what the tsType is
      // for enums

      if (
        column.enum ||
        column.type === 'json' ||
        column.type === 'jsonb' ||
        column.type === 'varying character'
      ) {
        fieldTemplates += niceString`
          @Field(() => ${graphQLDataType}, ${nullable})
          ${column.propertyName}${tsRequired}: ${tsType}
        `
      } else {
        fieldTemplates += niceString`
          @Field(${nullable})
          ${column.propertyName}${tsRequired}: ${tsType}
        `
      }

      fieldTemplates += '\n'
    }
  })

  return niceString`
    @InputType()
    export class ${entity.name}CreateInput {
    ${indent(fieldTemplates)}
    }
  `
}

// UPDATE

export const entityToUpdateInput = (entity: EntityMetadata): string => {
  let fieldTemplates = ''

  entity.columns.forEach((column: ColumnMetadata) => {
    if (
      !column.isPrimary &&
      !column.isCreateDate &&
      !column.isGenerated &&
      !column.isUpdateDate &&
      !column.isVersion &&
      !CREATE_UPDATE_BLACKLIST.includes(column.propertyName)
    ) {
      // TODO: also don't allow updated foreign key fields
      // Example: photo.userId: String
      const graphQLDataType = columnTypeToGraphQLDataType(column)
      const tsType = columnToTypeScriptType(column)

      if (
        column.enum ||
        column.type === 'json' ||
        column.type === 'jsonb' ||
        column.type === 'varying character'
      ) {
        fieldTemplates += niceString`
          @Field(() => ${graphQLDataType}, { nullable: true })
          ${column.propertyName}?: ${tsType}
        `
      } else {
        fieldTemplates += niceString`
          @Field({ nullable: true })
          ${column.propertyName}?: ${tsType}
        `
      }

      fieldTemplates += '\n'
    }
  })

  return niceString`
    @InputType()
    export class ${entity.name}UpdateInput {
    ${indent(fieldTemplates)}
    }
  `
}
