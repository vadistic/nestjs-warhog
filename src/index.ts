export * from './options'
export * from './factory'
export * from './common/find-operator'
export * from './common/order-condition'
export * from './generator/generate'
export * from './module/warthog.module'
export * from './module/warthog.service'
export * from './module/warthog.constants'
