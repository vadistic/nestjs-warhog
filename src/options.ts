import { Connection, ConnectionOptions } from 'typeorm'

export interface WarthogModuleOptions {
  /**
   * TODO:
   * Specify typeORM connection to use
   * (by default will use defualt TypeORM module connection).
   */
  connection?: Connection | ConnectionOptions | string
  /**
   * Entities glob. Needed to resolve entities import paths.
   */
  entities: string
  /**
   * Place for generated file. (Also needed to resolve paths for API use)
   */
  path: string
  /**
   * Generate base service classes
   */
  shouldGenerateServices?: boolean
  /**
   * Should prettier be used on generated files?
   */
  shouldSkipPrettier?: boolean
  /**
   * Custom disable logic (default is NODE_ENV !== 'development')
   */
  disabled?: boolean
}

export const defaultOptions: WarthogModuleOptions = {
  entities: '**/*.entity.ts',
  path: './generated/warthog.ts',
  shouldSkipPrettier: false,
  shouldGenerateServices: true,
  disabled: undefined,
}

export interface StringMap<T = any> {
  [key: string]: T
}
