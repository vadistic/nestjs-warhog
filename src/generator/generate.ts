import prettier from 'prettier'
import { EntityMetadata } from 'typeorm'

import {
  entityToWhereInput,
  entityToWhereUniqueInput,
  entityToCreateInput,
  entityToUpdateInput,
} from './entity-to-inputs'
import { entityToOrderByEnum } from './entity-to-enum'
import {
  entityToFindArgs,
  entityToCreateManyArgs,
  entityToUpdateArgs,
  entityToFindOneArgs,
  entityToCreateArgs,
  entityToDeleteManyArgs,
  entityToDeleteArgs,
  entityToUpdateManyArgs,
} from './entity-to-args'
import { WarthogModuleOptions } from '../options'

const HEADER = `
// This file has been auto-generated by NestJS Warthog.
// Do not update directly as it will be re-written.

import { ArgsType, Field, Float, InputType, Int } from 'type-graphql'
import { registerEnumType } from 'type-graphql'
import { PaginationArgs } from 'nestjs-warthog'
`

export const formatCode = (code: string) => {
  // TODO: grab our prettier options (single quote, etc...)
  try {
    return prettier.format(code, {
      parser: 'typescript',
      printWidth: 100,
      endOfLine: 'lf',
      singleQuote: true,
      semi: false,
      tabWidth: 2,
      useTabs: false,
    })
  } catch (err) {
    console.log(`There was a problem with formatting:`)
    console.log(JSON.stringify(err))

    return code
  }
}

export const generate = async (entities: EntityMetadata[], opts: WarthogModuleOptions) => {
  let result = HEADER.trim() + '\n\n'

  for (const entity of entities) {
    result += [
      entityToOrderByEnum(entity),
      entityToWhereInput(entity),
      entityToWhereUniqueInput(entity),
      entityToCreateInput(entity),
      entityToUpdateInput(entity),
      entityToFindOneArgs(entity),
      entityToFindArgs(entity),
      entityToCreateArgs(entity),
      entityToCreateManyArgs(entity),
      entityToUpdateArgs(entity),
      entityToUpdateManyArgs(entity),
      entityToDeleteArgs(entity),
      entityToDeleteManyArgs(entity),
    ].join('\n\n')
  }

  if (prettier && !opts.shouldSkipPrettier) {
    return formatCode(result)
  }

  return result
}
