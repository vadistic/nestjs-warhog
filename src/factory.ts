import { Connection } from 'typeorm'
import mkdirp from 'mkdirp'
import { promisify } from 'util'
import path from 'path'
import { writeFile } from 'fs'

import { WarthogModuleOptions, defaultOptions } from './options'
import { generate } from './generator/generate'

const mkdirP = promisify(mkdirp)
const writeFileP = promisify(writeFile)

/**
 * Allows generating inputs without utilising nestJS module
 *
 * API consistent with `@nestjs/graphql` `GraphQLDefinitionsFactory`
 */

export class WarthogDefinitionsFactory {
  constructor(readonly connection: Connection) {}

  async generate(opts?: WarthogModuleOptions) {
    const options = { ...defaultOptions, ...opts, connection: this.connection }

    await mkdirP(path.dirname(options.path))

    const definitions = await generate(this.connection.entityMetadatas, options)

    await writeFileP(options.path, definitions)
  }
}
