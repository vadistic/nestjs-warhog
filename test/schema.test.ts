import { ConnectionOptions, createConnection, Connection, EntityMetadata } from 'typeorm'

import { generate } from '../src/generator/generate'
import { WarthogModuleOptions } from '../src/options'
import { typeOrmOptions } from './fixture/connection'
import { Photo } from './fixture/photo.entity'
import Post from './fixture/post.entity'
import {
  entityToFindOneArgs,
  entityToFindArgs,
  entityToCreateArgs,
  entityToCreateManyArgs,
  entityToUpdateArgs,
  entityToUpdateManyArgs,
  entityToDeleteArgs,
  entityToDeleteManyArgs,
} from '../src/generator/entity-to-args'
import { entityToWhereUniqueInput, entityToWhereInput } from '../src/generator/entity-to-inputs'
import { entityToOrderByEnum } from '../src/generator/entity-to-enum'

export const options: WarthogModuleOptions = {
  path: './generated/warthog.ts',
  shouldSkipPrettier: false,
  disabled: undefined,
}

describe('schema', () => {
  let ctn: Connection
  let entity: EntityMetadata

  beforeAll(async () => {
    ctn = await createConnection({
      ...(typeOrmOptions as ConnectionOptions),
      entities: [Photo, Post],
    })

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    entity = ctn.entityMetadatas.find(val => val.name === 'Photo')!
  })

  afterAll(async () => {
    await ctn.close()
  })

  it('generate definitions that match snapshot', async () => {
    const schema = await generate(ctn.entityMetadatas, options)

    expect(schema).toMatchSnapshot()
  })

  it('generates OrderBy enum', () => {
    const res = entityToOrderByEnum(entity)

    expect(res).toMatchInlineSnapshot(`
      "export enum PhotoOrderByEnum {
        createdAt_ASC = 'createdAt_ASC',
        createdAt_DESC = 'createdAt_DESC',
        updatedAt_ASC = 'updatedAt_ASC',
        updatedAt_DESC = 'updatedAt_DESC',
        name_ASC = 'name_ASC',
        name_DESC = 'name_DESC',
        description_ASC = 'description_ASC',
        description_DESC = 'description_DESC',
        filename_ASC = 'filename_ASC',
        filename_DESC = 'filename_DESC',
        views_ASC = 'views_ASC',
        views_DESC = 'views_DESC',
        isPublished_ASC = 'isPublished_ASC',
        isPublished_DESC = 'isPublished_DESC',
        post_ASC = 'post_ASC',
        post_DESC = 'post_DESC',
      }

      registerEnumType(PhotoOrderByEnum, {
        name: 'PhotoOrderByInput'
      })"
    `)
  })

  it('generates WhereUnique input', () => {
    const res = entityToWhereUniqueInput(entity)

    expect(res).toMatchInlineSnapshot(`
      "@InputType()
      export class PhotoWhereUniqueInput {
        @Field(() => String)
        id?: string
      }"
    `)
  })

  it('generates Where input', () => {
    const res = entityToWhereInput(entity)

    expect(res).toMatchInlineSnapshot(`
      "@InputType()
      export class PhotoWhereInput {
        @Field(() => String,{ nullable: true })
        id_eq?: string
        @Field(() => [String], { nullable: true })
        id_in?: string[]
        
        @Field({ nullable: true })
        createdAt_gt?: Date
        @Field({ nullable: true })
        createdAt_gte?: Date
        @Field({ nullable: true })
        createdAt_lt?: Date
        @Field({ nullable: true })
        createdAt_lte?: Date
        
        @Field({ nullable: true })
        updatedAt_gt?: Date
        @Field({ nullable: true })
        updatedAt_gte?: Date
        @Field({ nullable: true })
        updatedAt_lt?: Date
        @Field({ nullable: true })
        updatedAt_lte?: Date
        
        @Field({ nullable: true })
        name_eq?: string
        @Field({ nullable: true })
        name_not?: string
        @Field({ nullable: true })
        name_contains?: string
        @Field({ nullable: true })
        name_startsWith?: string
        @Field({ nullable: true })
        name_endsWith?: string
        @Field(() => [String], { nullable: true })
        name_in?: string[]
        
        @Field({ nullable: true })
        description_eq?: string
        @Field({ nullable: true })
        description_not?: string
        @Field({ nullable: true })
        description_contains?: string
        @Field({ nullable: true })
        description_startsWith?: string
        @Field({ nullable: true })
        description_endsWith?: string
        @Field(() => [String], { nullable: true })
        description_in?: string[]
        
        @Field({ nullable: true })
        filename_eq?: string
        @Field({ nullable: true })
        filename_not?: string
        @Field({ nullable: true })
        filename_contains?: string
        @Field({ nullable: true })
        filename_startsWith?: string
        @Field({ nullable: true })
        filename_endsWith?: string
        @Field(() => [String], { nullable: true })
        filename_in?: string[]
        
        @Field({ nullable: true })
        views_eq?: number
        @Field({ nullable: true })
        views_gt?: number
        @Field({ nullable: true })
        views_gte?: number
        @Field({ nullable: true })
        views_lt?: number
        @Field({ nullable: true })
        views_lte?: number
        @Field(() => [Int], { nullable: true })
        views_in?: number[]
        
        @Field(() => Boolean, { nullable: true })
        isPublished_eq?: Boolean
        @Field(() => [Boolean], { nullable: true })
        isPublished_in?: Boolean[]
        
        @Field({ nullable: true })
        post_eq?: string
        @Field({ nullable: true })
        post_not?: string
        @Field({ nullable: true })
        post_contains?: string
        @Field({ nullable: true })
        post_startsWith?: string
        @Field({ nullable: true })
        post_endsWith?: string
        @Field(() => [String], { nullable: true })
        post_in?: string[]
      }"
    `)
  })

  it('generates findOne args', () => {
    const res = entityToFindOneArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoWhereUniqueArgs {
        @Field(() => PhotoWhereInput)
        where!: PhotoWhereUniqueInput
      }"
    `)
  })

  it('generates find args', () => {
    const res = entityToFindArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoWhereArgs extends PaginationArgs {
        @Field(() => PhotoWhereInput, { nullable: true })
        where?: PhotoWhereInput
        @Field(() => PhotoOrderByEnum, { nullable: true })
        orderBy?: PhotoOrderByEnum
      }"
    `)
  })

  it('generates create args', () => {
    const res = entityToCreateArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoCreateArgs {
        @Field(() => PhotoCreateInput)
        data!: PhotoCreateInput
      }"
    `)
  })

  it('generates createMany args', () => {
    const res = entityToCreateManyArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoCreateManyArgs {
        @Field(() => [PhotoCreateInput])
        data!: PhotoCreateInput[]
      }"
    `)
  })

  it('generates update args', () => {
    const res = entityToUpdateArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoUpdateArgs {
        @Field(() => PhotoUpdateInput)
        data!: PhotoUpdateInput
        @Field(() => PhotoWhereUniqueInput)
        where!: PhotoWhereUniqueInput
      }"
    `)
  })

  it('generates updateMany args', () => {
    const res = entityToUpdateManyArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoUpdateManyArgs {
        @Field(() => PhotoUpdateInput)
        data!: PhotoUpdateInput
        @Field(() => PhotoWhereInput)
        where!: PhotoWhereInput
      }"
    `)
  })

  it('generates delete args', () => {
    const res = entityToDeleteArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoDeleteArgs {
        @Field(() => PhotoWhereUniqueInput)
        where!: PhotoWhereUniqueInput
      }"
    `)
  })

  it('generates deleteMany args', () => {
    const res = entityToDeleteManyArgs(entity)

    expect(res).toMatchInlineSnapshot(`
      "@ArgsType()
      export class PhotoDeleteManyArgs {
        @Field(() => PhotoWhereInput)
        where!: PhotoWhereInput
      }"
    `)
  })
})
