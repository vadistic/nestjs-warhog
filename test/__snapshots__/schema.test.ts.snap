// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema generate definitions that match snapshot 1`] = `
"// This file has been auto-generated by NestJS Warthog.
// Do not update directly as it will be re-written.

import { ArgsType, Field, Float, InputType, Int } from 'type-graphql'
import { registerEnumType } from 'type-graphql'
import { PaginationArgs } from 'nestjs-warthog'

import { Photo } from '../fixture/photo.entity'
import Post from '../fixture/post.entity'

export enum PostOrderByEnum {
  title_ASC = 'title_ASC',
  title_DESC = 'title_DESC',
  text_ASC = 'text_ASC',
  text_DESC = 'text_DESC',
  likesCount_ASC = 'likesCount_ASC',
  likesCount_DESC = 'likesCount_DESC'
}

registerEnumType(PostOrderByEnum, {
  name: 'PostOrderByInput'
})

@InputType()
export class PostWhereInput {
  @Field(() => String, { nullable: true })
  id_eq?: string
  @Field(() => [String], { nullable: true })
  id_in?: string[]

  @Field({ nullable: true })
  title_eq?: string
  @Field({ nullable: true })
  title_not?: string
  @Field({ nullable: true })
  title_contains?: string
  @Field({ nullable: true })
  title_startsWith?: string
  @Field({ nullable: true })
  title_endsWith?: string
  @Field(() => [String], { nullable: true })
  title_in?: string[]

  @Field({ nullable: true })
  text_eq?: string
  @Field({ nullable: true })
  text_not?: string
  @Field({ nullable: true })
  text_contains?: string
  @Field({ nullable: true })
  text_startsWith?: string
  @Field({ nullable: true })
  text_endsWith?: string
  @Field(() => [String], { nullable: true })
  text_in?: string[]

  @Field({ nullable: true })
  likesCount_eq?: number
  @Field({ nullable: true })
  likesCount_gt?: number
  @Field({ nullable: true })
  likesCount_gte?: number
  @Field({ nullable: true })
  likesCount_lt?: number
  @Field({ nullable: true })
  likesCount_lte?: number
  @Field(() => [Float], { nullable: true })
  likesCount_in?: number[]
}

@InputType()
export class PostWhereUniqueInput {
  @Field(() => String)
  id?: string
}

@InputType()
export class PostCreateInput {
  @Field()
  title!: string
  @Field()
  text!: string
  @Field()
  likesCount!: number
}

@InputType()
export class PostUpdateInput {
  @Field({ nullable: true })
  title?: string
  @Field({ nullable: true })
  text?: string
  @Field({ nullable: true })
  likesCount?: number
}

@ArgsType()
export class PostWhereUniqueArgs {
  @Field(() => PostWhereInput)
  where!: PostWhereUniqueInput
}

@ArgsType()
export class PostWhereArgs extends PaginationArgs {
  @Field(() => PostWhereInput, { nullable: true })
  where?: PostWhereInput
  @Field(() => PostOrderByEnum, { nullable: true })
  orderBy?: PostOrderByEnum
}

@ArgsType()
export class PostCreateArgs {
  @Field(() => PostCreateInput)
  data!: PostCreateInput
}

@ArgsType()
export class PostCreateManyArgs {
  @Field(() => [PostCreateInput])
  data!: PostCreateInput[]
}

@ArgsType()
export class PostUpdateArgs {
  @Field(() => PostUpdateInput)
  data!: PostUpdateInput
  @Field(() => PostWhereUniqueInput)
  where!: PostWhereUniqueInput
}

@ArgsType()
export class PostUpdateManyArgs {
  @Field(() => PostUpdateInput)
  data!: PostUpdateInput
  @Field(() => PostWhereInput)
  where!: PostWhereInput
}

@ArgsType()
export class PostDeleteArgs {
  @Field(() => PostWhereUniqueInput)
  where!: PostWhereUniqueInput
}

@ArgsType()
export class PostDeleteManyArgs {
  @Field(() => PostWhereInput)
  where!: PostWhereInput
}
export enum PhotoOrderByEnum {
  createdAt_ASC = 'createdAt_ASC',
  createdAt_DESC = 'createdAt_DESC',
  updatedAt_ASC = 'updatedAt_ASC',
  updatedAt_DESC = 'updatedAt_DESC',
  name_ASC = 'name_ASC',
  name_DESC = 'name_DESC',
  description_ASC = 'description_ASC',
  description_DESC = 'description_DESC',
  filename_ASC = 'filename_ASC',
  filename_DESC = 'filename_DESC',
  views_ASC = 'views_ASC',
  views_DESC = 'views_DESC',
  isPublished_ASC = 'isPublished_ASC',
  isPublished_DESC = 'isPublished_DESC',
  post_ASC = 'post_ASC',
  post_DESC = 'post_DESC'
}

registerEnumType(PhotoOrderByEnum, {
  name: 'PhotoOrderByInput'
})

@InputType()
export class PhotoWhereInput {
  @Field(() => String, { nullable: true })
  id_eq?: string
  @Field(() => [String], { nullable: true })
  id_in?: string[]

  @Field({ nullable: true })
  createdAt_gt?: Date
  @Field({ nullable: true })
  createdAt_gte?: Date
  @Field({ nullable: true })
  createdAt_lt?: Date
  @Field({ nullable: true })
  createdAt_lte?: Date

  @Field({ nullable: true })
  updatedAt_gt?: Date
  @Field({ nullable: true })
  updatedAt_gte?: Date
  @Field({ nullable: true })
  updatedAt_lt?: Date
  @Field({ nullable: true })
  updatedAt_lte?: Date

  @Field({ nullable: true })
  name_eq?: string
  @Field({ nullable: true })
  name_not?: string
  @Field({ nullable: true })
  name_contains?: string
  @Field({ nullable: true })
  name_startsWith?: string
  @Field({ nullable: true })
  name_endsWith?: string
  @Field(() => [String], { nullable: true })
  name_in?: string[]

  @Field({ nullable: true })
  description_eq?: string
  @Field({ nullable: true })
  description_not?: string
  @Field({ nullable: true })
  description_contains?: string
  @Field({ nullable: true })
  description_startsWith?: string
  @Field({ nullable: true })
  description_endsWith?: string
  @Field(() => [String], { nullable: true })
  description_in?: string[]

  @Field({ nullable: true })
  filename_eq?: string
  @Field({ nullable: true })
  filename_not?: string
  @Field({ nullable: true })
  filename_contains?: string
  @Field({ nullable: true })
  filename_startsWith?: string
  @Field({ nullable: true })
  filename_endsWith?: string
  @Field(() => [String], { nullable: true })
  filename_in?: string[]

  @Field({ nullable: true })
  views_eq?: number
  @Field({ nullable: true })
  views_gt?: number
  @Field({ nullable: true })
  views_gte?: number
  @Field({ nullable: true })
  views_lt?: number
  @Field({ nullable: true })
  views_lte?: number
  @Field(() => [Int], { nullable: true })
  views_in?: number[]

  @Field(() => Boolean, { nullable: true })
  isPublished_eq?: Boolean
  @Field(() => [Boolean], { nullable: true })
  isPublished_in?: Boolean[]

  @Field({ nullable: true })
  post_eq?: string
  @Field({ nullable: true })
  post_not?: string
  @Field({ nullable: true })
  post_contains?: string
  @Field({ nullable: true })
  post_startsWith?: string
  @Field({ nullable: true })
  post_endsWith?: string
  @Field(() => [String], { nullable: true })
  post_in?: string[]
}

@InputType()
export class PhotoWhereUniqueInput {
  @Field(() => String)
  id?: string
}

@InputType()
export class PhotoCreateInput {
  @Field()
  name!: string
  @Field()
  description!: string
  @Field()
  filename!: string
  @Field()
  views!: number
  @Field()
  isPublished!: boolean
  @Field({ nullable: true })
  post?: string
}

@InputType()
export class PhotoUpdateInput {
  @Field({ nullable: true })
  name?: string
  @Field({ nullable: true })
  description?: string
  @Field({ nullable: true })
  filename?: string
  @Field({ nullable: true })
  views?: number
  @Field({ nullable: true })
  isPublished?: boolean
  @Field({ nullable: true })
  post?: string
}

@ArgsType()
export class PhotoWhereUniqueArgs {
  @Field(() => PhotoWhereInput)
  where!: PhotoWhereUniqueInput
}

@ArgsType()
export class PhotoWhereArgs extends PaginationArgs {
  @Field(() => PhotoWhereInput, { nullable: true })
  where?: PhotoWhereInput
  @Field(() => PhotoOrderByEnum, { nullable: true })
  orderBy?: PhotoOrderByEnum
}

@ArgsType()
export class PhotoCreateArgs {
  @Field(() => PhotoCreateInput)
  data!: PhotoCreateInput
}

@ArgsType()
export class PhotoCreateManyArgs {
  @Field(() => [PhotoCreateInput])
  data!: PhotoCreateInput[]
}

@ArgsType()
export class PhotoUpdateArgs {
  @Field(() => PhotoUpdateInput)
  data!: PhotoUpdateInput
  @Field(() => PhotoWhereUniqueInput)
  where!: PhotoWhereUniqueInput
}

@ArgsType()
export class PhotoUpdateManyArgs {
  @Field(() => PhotoUpdateInput)
  data!: PhotoUpdateInput
  @Field(() => PhotoWhereInput)
  where!: PhotoWhereInput
}

@ArgsType()
export class PhotoDeleteArgs {
  @Field(() => PhotoWhereUniqueInput)
  where!: PhotoWhereUniqueInput
}

@ArgsType()
export class PhotoDeleteManyArgs {
  @Field(() => PhotoWhereInput)
  where!: PhotoWhereInput
}
"
`;
